File: Android/app/src/main/java/com/geniapassword/mobile/security/PasswordGenerator.kt
package com.geniapassword.mobile.security

import java.security.SecureRandom

object PasswordGenerator {
    private val random = SecureRandom()

    private const val LOWER = "abcdefghijklmnopqrstuvwxyz"
    private const val UPPER = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    private const val DIGITS = "0123456789"
    private const val SYMBOLS = "!@#$%^&*()-_=+[]{}|;:,.<>?"
    private const val AMBIGUOUS = "O0oIl1|`'\""

    fun generate(
        length: Int = 16,
        useUpper: Boolean = true,
        useDigits: Boolean = true,
        useSymbols: Boolean = true,
        excludeAmbiguous: Boolean = false,
    ): String {
        require(length in 8..128) { "Длина пароля должна быть от 8 до 128 символов." }

        fun pool(value: String): String =
            if (!excludeAmbiguous) value else value.filterNot { it in AMBIGUOUS }

        val requiredPools = buildList {
            add(pool(LOWER))
            if (useUpper) add(pool(UPPER))
            if (useDigits) add(pool(DIGITS))
            if (useSymbols) add(pool(SYMBOLS))
        }.filter { it.isNotEmpty() }

        require(requiredPools.isNotEmpty()) { "Выберите хотя бы одну группу символов." }
        require(length >= requiredPools.size) { "Длина пароля слишком мала для выбранных групп символов." }

        val fullPool = requiredPools.joinToString(separator = "")
        val chars = CharArray(length)
        var index = 0

        requiredPools.forEach { currentPool ->
            chars[index++] = currentPool[random.nextInt(currentPool.length)]
        }
        while (index < chars.size) {
            chars[index++] = fullPool[random.nextInt(fullPool.length)]
        }

        // Fisher-Yates с SecureRandom: обязательные символы не остаются в начале.
        for (i in chars.lastIndex downTo 1) {
            val j = random.nextInt(i + 1)
            val tmp = chars[i]
            chars[i] = chars[j]
            chars[j] = tmp
        }
        return String(chars)
    }
}

File: Windows/PasswordNotebook/Security/PasswordGenerator.cs
using System.Security.Cryptography;

namespace GeniaPassword.Security;

internal static class PasswordGenerator
{
    private const string Lowercase = "abcdefghijklmnopqrstuvwxyz";
    private const string Uppercase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    private const string Digits = "0123456789";
    private const string Symbols = "!@#$%^&*()-_=+[]{}|;:,.<>?";
    private const string Ambiguous = "O0oIl1|`'\"";

    internal static string Generate(int length = 16) =>
        Generate(length, useUpper: true, useDigits: true, useSymbols: true, excludeAmbiguous: false);

    internal static string Generate(
        int length,
        bool useUpper,
        bool useDigits,
        bool useSymbols,
        bool excludeAmbiguous)
    {
        if (length is < 8 or > 128)
        {
            throw new ArgumentOutOfRangeException(nameof(length), "Длина пароля должна быть от 8 до 128 символов.");
        }

        string lower = Filter(Lowercase, excludeAmbiguous);
        string upper = Filter(Uppercase, excludeAmbiguous);
        string digits = Filter(Digits, excludeAmbiguous);
        string symbols = Filter(Symbols, excludeAmbiguous);

        var pools = new List<string> { lower };
        if (useUpper)
        {
            pools.Add(upper);
        }
        if (useDigits)
        {
            pools.Add(digits);
        }
        if (useSymbols)
        {
            pools.Add(symbols);
        }

        pools.RemoveAll(static pool => pool.Length == 0);
        if (pools.Count == 0)
        {
            throw new InvalidOperationException("Выберите хотя бы одну группу символов.");
        }
        if (length < pools.Count)
        {
            throw new InvalidOperationException("Длина пароля слишком мала для выбранных групп символов.");
        }

        string allChars = string.Concat(pools);
        char[] password = new char[length];
        int index = 0;

        foreach (string pool in pools)
        {
            password[index++] = GetRandomChar(pool);
        }

        while (index < password.Length)
        {
            password[index++] = GetRandomChar(allChars);
        }

        Shuffle(password);
        string result = new(password);
        Array.Clear(password, 0, password.Length);
        return result;
    }

    private static string Filter(string source, bool excludeAmbiguous) =>
        excludeAmbiguous
            ? new string(source.Where(character => !Ambiguous.Contains(character)).ToArray())
            : source;

    private static char GetRandomChar(string characterSet)
    {
        int index = RandomNumberGenerator.GetInt32(characterSet.Length);
        return characterSet[index];
    }

    private static void Shuffle(char[] array)
    {
        for (int i = array.Length - 1; i > 0; i--)
        {
            int j = RandomNumberGenerator.GetInt32(i + 1);
            (array[i], array[j]) = (array[j], array[i]);
        }
    }
}
